File: BOJ/자료구조/9012.py
# 9012 괄호
# https://www.acmicpc.net/problem/9012

def solution(parentheses):
    stack = []

    for p in parentheses:
        if p == '(':
            stack.append(0)
        else:
            if not len(stack):
                return 'NO'
            stack.pop()

    return 'YES' if not len(stack) else 'NO'


n = int(input())

for _ in range(n):
    print(solution(input()))

File: BOJ/자료구조/9012/9012.cpp
// 9012 괄호
// https://www.acmicpc.net/problem/9012
#include <iostream>
#include <vector>

using namespace std;

string solution(string s) {
    string answer = "YES";
    vector<int> stack;

    for (int i = 0; i < s.size(); ++i)
    {
        if (s[i] == '(')
        {
            stack.push_back(1);
        } else {
            if (stack.empty()) 
            {
                answer = "NO";
                break;
            }
            stack.pop_back();
        }
    }

    if (!stack.empty()) 
    {
        answer = "NO";
    }

    return answer;
}

int main(void) {
    int T;
    string s;

    cin >> T;
    while (T--)
    {
        cin >> s;
        cout << solution(s) << '\n';
    }
    

    return 0;
}